### PR TITLE
rename gcc-wrapper to cc-wrapper and make compatible with clang

### PR DESCRIPTION
**Pull request by gridaphobe**
This is from the darwin-stdenv branch and should be the minimum changes required to make our stdenv usable with the unified cc-wrapper.

/cc @shlevy @jwiegley @joelteon @copumpkin

### CODE REVIEW

**Comment by jwiegley** [comment]
+1

**Comment by shlevy** [comment]
I'll create a hydra jobset for this later today.

**Comment by gridaphobe** [comment]
Ah, this totally breaks the 
```
assert stdenv.cc ? gcc;
```
assertions in glibc and glib. What would be the best way to fix them, perhaps expose the `ccName` and write
```
assert stdenv.cc.ccName == "gcc";
```
?

**Comment by shlevy** [comment]
We should probably instead consider what those checks should actually mean/whether they are necessary. E.g.: Why does glibc require gcc? Is it instead that glibc requires not darwin?

**Comment by gridaphobe** [comment]
It looks like the only packages that have this type of assertion are `glibc`, `glib`, `guile-lib`, and `gcl`.

I think glibc uses gcc-specific extensions, but I don't know about the others.

**Comment by gridaphobe** [comment]
(See https://sourceware.org/ml/libc-alpha/2014-05/msg00707.html)

**Comment by shlevy** [comment]
http://hydra.nixos.org/jobset/nixpkgs/pr-5676

File: pkgs/build-support/cc-wrapper/cc-wrapper.sh [hunk 1]
```diff
@@ -79,6 +79,18 @@ if [ "$NIX_ENFORCE_PURITY" = 1 -a -n "$NIX_STORE" ]; then
     params=("${rest[@]}")
 fi
 
+if test -n "@libcxx@"; then
+    NIX_CFLAGS_COMPILE="$NIX_CFLAGS_COMPILE -isystem@libcxx@/include/c++/v1"
```
Comment (edolstra) [inline pkgs/build-support/cc-wrapper/cc-wrapper.sh]: Why is this needed? Libc++ already has a setup hook (`pkgs/development/libraries/libc++/setup-hook.sh`) that sets these flags:
```
export NIX_CFLAGS_COMPILE+=" -isystem @out@/include/c++/v1 -stdlib=libc++"
export NIX_CFLAGS_LINK+=" -stdlib=libc++ -Wl,-rpath,@libcxxabi@/lib"
```
I introduced that on purpose to simplify the (g)cc wrapper.

Comment (gridaphobe) [inline pkgs/build-support/cc-wrapper/cc-wrapper.sh]: Perhaps @copumpkin or @joelteon can comment? I didn't add this, but I know there was some trickiness with `libc++` that required the extra `NIX_SKIP_CXX[ABI]` variables. Though I suppose they could be merged into the `libc++` derivation.

Comment (shlevy) [inline pkgs/build-support/cc-wrapper/cc-wrapper.sh]: Yes, let's just put the `NIX_SKIP_CXX[ABI]` variables into the setup hook as well. @gridaphobe If that's obvious how you should do it please do, otherwise I'll take a look when I get a chance.

File: pkgs/build-support/cc-wrapper/cc-wrapper.sh [hunk 2]
```diff
@@ -79,6 +79,18 @@ if [ "$NIX_ENFORCE_PURITY" = 1 -a -n "$NIX_STORE" ]; then
...
+    if [[ "@prog@" = *++ ]]; then
+        NIX_CFLAGS_COMPILE="$NIX_CFLAGS_COMPILE -stdlib=libc++"
+        if test -z "$NIX_SKIP_CXX"; then
+            NIX_CFLAGS_LINK="$NIX_CFLAGS_LINK -L@libcxx@/lib -stdlib=libc++"
+        fi
+        if test -z "$NIX_SKIP_CXXABI" && echo "$@" | grep -qvw -- -nostdlib; then
```
Comment (shlevy) [inline pkgs/build-support/cc-wrapper/cc-wrapper.sh]: @edolstra Any idea on how to move this into the setup hook?

**Comment by vcunat** [comment]
I pushed this to staging; maybe it was a bit early, but tarball and some other things work OK (I didn't notice the pr-5676 jobset started to evaluate).

**Comment by shlevy** [comment]
@vcunat Ah, OK. I was planning to just merge to master once hydra finished, since (unlike something like a gcc upgrade) this change should ideally produce identical results to what we had working before. But if it's already in staging, great! I'll keep the jobset running to isolate issues due to this change, but I'll close this.

**Comment by vcunat** [comment]
Hmm, I spot a build problem most likely due to this: http://hydra.nixos.org/build/18936280 (nothing else in the range was playing with compiler options).

**Comment by shlevy** [comment]
Yup, can reproduce on this branch. Will look for a fix.

**Comment by vcunat** [comment]
It seems this PR brings no significant problems. I looked at a dozen "failures" from http://hydra.nixos.org/eval/1171905. For all of them, either they fail on master or I can't reproduce them, at least on later staging cfb29ab88 (many of those are because of qt5 fail, but that works for me).

That jobset is probably no longer needed. Otherwise it will likely continue building new things, as it follows @gridaphobe's master.

**Comment by shlevy** [comment]
@edolstra @rbvermaa  tried to delete the jobset to no effect, can you guys nuke it in the db?

**Comment by edolstra** [comment]
You shouldn't delete jobsets, just disable/hide them.

File: pkgs/build-support/cc-wrapper/cc-wrapper.sh (hunk 1 above)
Comment (copumpkin) [inline pkgs/build-support/cc-wrapper/cc-wrapper.sh]: @edolstra I think that behavior is different from what we had before, since now every call to your `cc` (not `c++`) results in you getting `-stdlib=libc++` passed in. Before, this was conditional on whether your invocation was to `*++`. Could that lead to bad effects? We could obviously just make up a `NIX_CONDITIONAL_CXX_LDFLAGS` or something and set that in the libc++ setup hook.

I actually like this approach far better, because then we don't even need the skip variables anymore. We just don't have libc++abi depend on libc++abi. Although if we have the stdenv forces us to depend on libc++ then that might get tricky, unless there's an easy way to opt out of build inputs that the stdenv gives you.

Comment (copumpkin) [inline pkgs/build-support/cc-wrapper/cc-wrapper.sh]: @edolstra I also just noticed that your hook doesn't explicitly link with `-lc++abi`. Was that intentional?
